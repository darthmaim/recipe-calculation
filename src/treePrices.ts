// Update the tree prices
export default function treePrices(tree: any, itemPrices: any) {
  tree = { ...tree }

  // Calculate the buy prices
  tree.buyPriceEach = itemPrices[tree.id] || false
  tree.buyPrice = tree.buyPriceEach ? tree.usedQuantity * tree.buyPriceEach : false
  tree.decisionPrice = tree.buyPrice

  if (!tree.components) {
    return tree
  }

  // Calculate the tree prices traversal for the sub-components
  tree.components = tree.components.map((component: any) => treePrices(component, itemPrices))

  // Calculate the craft price out of the best prices
  tree.craftPrice = tree.components.map((c: any) => c.decisionPrice).reduce((a: any, b: any) => a + b)

  // If we explicitly don't craft this, keep the buy price as the best price
  if (tree.craft === false) {
    return tree
  }

  // Update the decision price of this tree segment to the craft price,
  // used to determine the craft price of the higher-up recipe
  if (tree.craft === true || !tree.buyPrice || tree.craftPrice < tree.buyPrice) {
    tree.decisionPrice = tree.craftPrice
  }

  return tree
}
