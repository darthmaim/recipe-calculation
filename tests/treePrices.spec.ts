// @ts-nocheck
import treePrices from '../src/treePrices'

describe('treePrices', () => {
  it('updates the prices for a tree', () => {
    const recipeTree = {
      id: 1,
      usedQuantity: 1,
      components: [
        { id: 2, usedQuantity: 1 },
        { id: 3, usedQuantity: 2 },
      ],
    }
    const prices = { 1: 123, 2: 1, 3: 2 }

    const calculatedTree = treePrices(recipeTree, prices)
    expect(calculatedTree).toMatchSnapshot()
  })

  it('updates the prices for a tree with flags correctly', () => {
    const recipeTree = {
      id: 1,
      usedQuantity: 1,
      craft: false,
      components: [
        { id: 2, usedQuantity: 1 },
        { id: 3, usedQuantity: 2 },
      ],
    }
    const prices = { 1: 123, 2: 1, 3: 2 }

    const calculatedTree = treePrices(recipeTree, prices)
    expect(calculatedTree).toMatchSnapshot()
  })

  it('updates the prices for a tree with missing buy prices', () => {
    const recipeTree = {
      id: 1,
      usedQuantity: 1,
      components: [
        { id: 2, usedQuantity: 1 },
        {
          id: 3,
          usedQuantity: 2,
          components: [
            { id: 4, usedQuantity: 50 },
            { id: 5, usedQuantity: 2 },
          ],
        },
      ],
    }
    const prices = { 1: 123, 2: 1, 3: 2, 5: 10 }

    const calculatedTree = treePrices(recipeTree, prices)
    expect(calculatedTree).toMatchSnapshot()
  })
})
