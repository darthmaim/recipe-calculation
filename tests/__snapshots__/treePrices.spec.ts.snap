// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`treePrices updates the prices for a tree 1`] = `
Object {
  "buyPrice": 123,
  "buyPriceEach": 123,
  "components": Array [
    Object {
      "buyPrice": 1,
      "buyPriceEach": 1,
      "components": undefined,
      "decisionPrice": 1,
      "id": 2,
      "usedQuantity": 1,
    },
    Object {
      "buyPrice": 4,
      "buyPriceEach": 2,
      "components": undefined,
      "decisionPrice": 4,
      "id": 3,
      "usedQuantity": 2,
    },
  ],
  "craftPrice": 5,
  "decisionPrice": 5,
  "id": 1,
  "usedQuantity": 1,
}
`;

exports[`treePrices updates the prices for a tree with flags correctly 1`] = `
Object {
  "buyPrice": 123,
  "buyPriceEach": 123,
  "components": Array [
    Object {
      "buyPrice": 1,
      "buyPriceEach": 1,
      "components": undefined,
      "decisionPrice": 1,
      "id": 2,
      "usedQuantity": 1,
    },
    Object {
      "buyPrice": 4,
      "buyPriceEach": 2,
      "components": undefined,
      "decisionPrice": 4,
      "id": 3,
      "usedQuantity": 2,
    },
  ],
  "craft": false,
  "craftPrice": 5,
  "decisionPrice": 123,
  "id": 1,
  "usedQuantity": 1,
}
`;

exports[`treePrices updates the prices for a tree with missing buy prices 1`] = `
Object {
  "buyPrice": 123,
  "buyPriceEach": 123,
  "components": Array [
    Object {
      "buyPrice": 1,
      "buyPriceEach": 1,
      "components": undefined,
      "decisionPrice": 1,
      "id": 2,
      "usedQuantity": 1,
    },
    Object {
      "buyPrice": 4,
      "buyPriceEach": 2,
      "components": Array [
        Object {
          "buyPrice": false,
          "buyPriceEach": false,
          "components": undefined,
          "decisionPrice": false,
          "id": 4,
          "usedQuantity": 50,
        },
        Object {
          "buyPrice": 20,
          "buyPriceEach": 10,
          "components": undefined,
          "decisionPrice": 20,
          "id": 5,
          "usedQuantity": 2,
        },
      ],
      "craftPrice": 20,
      "decisionPrice": 4,
      "id": 3,
      "usedQuantity": 2,
    },
  ],
  "craftPrice": 5,
  "decisionPrice": 5,
  "id": 1,
  "usedQuantity": 1,
}
`;
